### PR TITLE
Feature: Fullstack music licensing workflow challenge implementation

### PR DESCRIPTION
**Pull request by AlanGarcilazo**
This pull request includes the complete fullstack solution for the music licensing workflow challenge.

The implementation is fully modularized:
- Backend setup with NestJS, Docker, database models, and seeding.
- Frontend setup with Next.js, custom hooks, service layer, and reusable pages.
- Static assets and global styling.
- Full technical documentation in the `technical-decisions` folder, including architecture diagrams and commit strategy.

Each commit was structured to keep changes atomic, focused, and easy to review.

Thanks for reviewing! 🚀

### CODE REVIEW

File: backend/src/app.module.ts
Comment (ulises-jeremias) [inline backend/src/app.module.ts:1]: Good modularization using NestJS modules! 🔥 

An improvement could be to add global exception filters and validation pipes for better error handling (see [NestJS docs](https://docs.nestjs.com/exception-filters)).

File: backend/src/app.module.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import { SongModule } from './song/song.module';
+
+@Module({
+  imports: [
+    TypeOrmModule.forRoot({
+      type: 'postgres',
+      host: 'postgres',
+      port: 5432,
+      username: 'postgres',
+      password: 'postgres',
```
Comment (ulises-jeremias) [inline backend/src/app.module.ts:21]: Another improvement could be to use environment variables for configuration (e.g., database connection), not hardcoded values

File: backend/src/entities/movie.entity.ts
Comment (ulises-jeremias) [inline backend/src/entities/movie.entity.ts:1]: Entity relationships are well defined 🥳 

Some ways to Improve entities:
- Add validation decorators (e.g., @IsString, @IsNotEmpty) for DTOs.
- Add `createdAt` and `updatedAt` timestamps for auditability.
- Consider using UUIDs for primary keys for better scalability.

File: backend/src/entities/scene.entity.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  id: number;
+
+  @Column()
+  name: string;
+
+  @ManyToOne(() => Movie, (movie) => movie.scenes)
+  movie: Movie;
+
+  @OneToMany(() => Track, (track) => track.scene)
+  tracks: Track[];
```
Comment (ulises-jeremias) [inline backend/src/entities/scene.entity.ts:23]: Correct use of relations to Movie and Track 🔥

Comment (ulises-jeremias) [reply backend/src/entities/scene.entity.ts:23]: Something to consider is to add cascade options if you want scenes to be deleted when a movie is deleted. Also to add index on foreign keys for performance 😊

File: backend/src/entities/track.entity.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  id: number;
+
+  @Column()
+  startTime: string;
+
+  @Column()
+  endTime: string;
+
+  @ManyToOne(() => LicenseStatus)
+  licenseStatus: LicenseStatus;
```
Comment (ulises-jeremias) [inline backend/src/entities/track.entity.ts:25]: Licensing status is modeled as an enum! 🔥 Great idea!

File: backend/src/entities/track.entity.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+@Entity()
+export class Track {
+  @PrimaryGeneratedColumn()
+  id: number;
+
+  @Column()
+  startTime: string;
+
+  @Column()
+  endTime: string;
```
Comment (ulises-jeremias) [inline backend/src/entities/track.entity.ts:22]: An improvement could be to add validation for start/end times (e.g., end > start) 👀

File: backend/src/entities/track.entity.ts (hunk 3 above)
Comment (ulises-jeremias) [reply backend/src/entities/track.entity.ts:25]: Consider a separate table for licensing history (audit trail) 😊

File: backend/src/entities/song.entity.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import { Entity, PrimaryGeneratedColumn, Column, OneToOne } from 'typeorm';
+import { Track } from './track.entity';
+
+@Entity()
+export class Song {
```
Comment (ulises-jeremias) [inline backend/src/entities/song.entity.ts:5]: Song entity is simple and clear! 😊

File: backend/src/entities/song.entity.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  @PrimaryGeneratedColumn()
+  id: number;
+
+  @Column()
+  title: string;
+
+  @Column()
+  artist: string;
```
Comment (ulises-jeremias) [inline backend/src/entities/song.entity.ts:13]: Something to improve this definition could be to add unique constraint on song title + artist to avoid duplicates

File: backend/src/track/track.controller.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+import { TrackService } from './track.service';
+import { Track } from '../entities/track.entity';
+import { ParseIntPipe } from '@nestjs/common';
+
+@Controller('track')
+export class TrackController {
+  constructor(private readonly trackService: TrackService) {}
+
+  @Get()
+  findAll(@Query('sceneId') sceneId?: string): Promise<Track[]> {
```
Comment (ulises-jeremias) [inline backend/src/track/track.controller.ts:11]: 🔥

File: backend/src/track/track.controller.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    }
+    return this.trackService.findAll();
+  }
+  @Delete(':id')
+  async remove(@Param('id', ParseIntPipe) id: number): Promise<void> {
+  await this.trackService.remove(id); 
+  }
+
+  @Post()
+  create(@Body() data: any): Promise<Track> {
```
Comment (ulises-jeremias) [inline backend/src/track/track.controller.ts:23]: Using a type or interface and a typeguard to validate the shape of data would be ideal in here!

File: backend/src/track/track.controller.ts
Comment (ulises-jeremias) [inline backend/src/track/track.controller.ts:1]: Improvement for Controllers:
- Add error handling for not found and invalid input.
- Use HTTP status codes consistently (e.g., 201 for creation).
- Add authentication/authorization middleware for production.

File: backend/src/track/track.service.ts
Comment (ulises-jeremias) [inline backend/src/track/track.service.ts:1]: The business logic is well separated from controllers 😊

Comment (ulises-jeremias) [reply backend/src/track/track.service.ts:1]: Some improvements for services could be:

- Add transaction management for multi-step operations.
- Handle edge cases (e.g., updating non-existent track).
- Add logging for important actions

File: backend/src/license-status/license-status.gateway.ts [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+import { LicenseStatus } from '../entities/license-status.entity';
+
+@WebSocketGateway({
+  cors: {
+    origin: '*',
+  },
+})
+export class LicenseStatusGateway {
+  @WebSocketServer()
+  server: Server;
```
Comment (ulises-jeremias) [inline backend/src/license-status/license-status.gateway.ts:12]: WebSocket gateway is implemented for real-time updates! Awesome! 🔥

Comment (ulises-jeremias) [reply backend/src/license-status/license-status.gateway.ts:12]: Some improvements could be:

- Add authentication to WebSocket connections.
- Handle disconnects and errors gracefully.
- Document event payloads.

File: backend/src/license-status/license-status.gateway.spec.ts
Comment (ulises-jeremias) [inline backend/src/license-status/license-status.gateway.spec.ts:1]: Great to know you added unit tests!

Some things to consider:
- Use test data factories to avoid duplication.
- Increase test coverage, especially for edge cases and error handling.

File: frontend/src/hooks/useScenes.ts [hunk 10]
```diff
@@ -0,0 +1,31 @@
+import { useEffect, useState } from 'react';
+import axios from 'axios';
+
+type Scene = {
+  id: number;
+  name: string;
+};
+
+export function useScenes(movieId: string | number | undefined) {
```
Comment (ulises-jeremias) [inline frontend/src/hooks/useScenes.ts:9]: Great that you used custom hooks for fetching data!

Some improvements to this hooks could be:
- Add error state management
- Use SWR or React Query for caching, error handling and revalidation
- Add types for returned data

File: frontend/src/services/movieService.ts
Comment (ulises-jeremias) [inline frontend/src/services/movieService.ts:1]: What’s Good:
- Uses axios, which simplifies HTTP requests and response parsing.
- Function is asynchronous and returns the data directly.

Areas for Improvement:
- Hardcoded URL: The API URL is hardcoded. For production, always use environment variables (e.g., process.env.NEXT_PUBLIC_API_BASE_URL) to allow flexibility between environments.
- Error Handling: There is no error handling. If the request fails, the error will propagate unhandled. Wrap the call in a try/catch and handle or rethrow errors with context.
- TypeScript Typing: The function does not specify a return type. For maintainability and type safety, define and use a Movie type/interface.
- Reusability: If you have multiple endpoints, consider creating an axios instance with a base URL and common config.
- Naming Consistency: The endpoint is /movie, but REST conventions usually use /movies for collections.

Example refactor:

```ts
import axios from 'axios';
import type { Movie } from '../types/movie'; // Define this interface in your types

const API_BASE = process.env.NEXT_PUBLIC_API_BASE_URL || 'http://localhost:3000';

export async function getMovies(): Promise<Movie[]> {
  try {
    const res = await axios.get(`${API_BASE}/movies`);
    return res.data;
  } catch (error) {
    // Optionally log error or transform it for UI
    throw new Error('Failed to fetch movies');
  }
}
```

File: backend/Dockerfile
Comment (ulises-jeremias) [inline backend/Dockerfile:1]: Great you used docker!

The container runs npm run start:dev, which is intended for development (hot reload, less performance, more memory). For production, use npm run build and npm run start.

**Suggestion:** Use multi-stage builds to separate build and runtime, and run in production mode for deployment.

Comment (ulises-jeremias) [reply backend/Dockerfile:1]: Example:

```Dockerfile
FROM node:18 AS builder
WORKDIR /app
COPY package*.json ./
RUN npm install
COPY . .
RUN npm run build

FROM node:18
WORKDIR /app
COPY --from=builder /app .
ENV NODE_ENV=production
EXPOSE 3000
CMD ["npm", "run", "start"]
```

File: docker-compose.yml
Comment (ulises-jeremias) [inline docker-compose.yml:1]: Nice compose file! Just some comments to keep learning!

- The frontend (Next.js) is not dockerized, so the full stack cannot be run with a single docker-compose up. **Suggestion:** Add a Dockerfile for the frontend and include it as a service in docker-compose.yml.
- docker-compose.yml should define environment variables for each service and use a .env file for secrets/configuration. **Suggestion:** Add a .env.example and reference it in docker-compose.yml.
- No healthcheck is defined for the backend service in Docker Compose, which is useful for orchestration and monitoring.

File: technical-decisions/README_DETAILS.md
Comment (ulises-jeremias) [inline technical-decisions/README_DETAILS.md:1]: Your technical documentation is clear, comprehensive, and well-structured. It covers the tech stack, project structure, API endpoints, setup, and technical decisions, which is excellent for maintainability and onboarding.
However, it’s a common convention to store this type of documentation in a /docs directory at the root of the repository, making it easier for collaborators to find and reference. Also, consider splitting specific topics (like architecture diagrams, API reference, or setup guides) into separate markdown files within /docs for even better organization.
Additionally, you might want to leverage standard files such as CHANGELOG.md for tracking changes, CONTRIBUTING.md for contribution guidelines, and ARCHITECTURE.md for system design details, depending on what you want to communicate.

File: technical-decisions/README_Commits.md
Comment (ulises-jeremias) [inline technical-decisions/README_Commits.md:1]: The commit strategy overview is a great touch—it demonstrates a professional workflow and helps reviewers understand the project’s evolution. For even better discoverability, consider moving this file (and similar process or workflow docs) into a /docs directory.
If you want to document release history or major changes, you can use a CHANGELOG.md at the root. For documenting your workflow or conventions, a CONTRIBUTING.md or a dedicated section in the main README.md can also be helpful.

**Review by ulises-jeremias — APPROVED** [review]
Hey @AlanGarcilazo – congrats on submitting your PR! 🎉

This is a solid solution that shows a lot of care and good thinking. A few highlights worth mentioning:
- **Clear and organized commits**: Your commit history is clean and descriptive, which really helps when reviewing and following the project’s evolution.
- **Good structure**: The separation between backend, frontend, and data layers is well thought out and aligned with good practices.
- **Relevant tech choices**: You used a modern and appropriate stack (NestJS, Next.js, PostgreSQL, Docker, WebSockets), and you covered all the main requirements—including real-time updates.
- **Helpful documentation**: The notes and explanations you included are easy to follow and show good communication skills, especially at a junior level.

I've left some suggestions and comments that point out areas for improvement -these are normal and expected, and meant to help you grow. Overall, this is a strong submission that reflects both technical skills and a thoughtful approach. Nice job! 🚀

**Comment by ulises-jeremias** [comment]
@AlanGarcilazo things to consider for auth:

- Selfhosted auth: https://www.passportjs.org/
- Using Auth providers:
  - https://clerk.com/
  - https://auth0.com/

**Comment by AlanGarcilazo** [comment]
Buenas tardes Ulises, como estas?

Quería contarte que estuve revisando los puntos de mejora que me
compartiste y voy a estar subiendo esos cambios entre esta semana y la
próxima. El camino se me demoró un poco porque estuve con algunas
prioridades de proyectos personales que necesitaba cerrar y, además, me
tomé un tiempo para capacitarme mejor en la tecnología antes de meter los
ajustes.

De todas formas, ya tengo todo encaminado y en breve pusheo los updates!

Gracias por la paciencia y la buena onda! Saludos 😁
